crates/ra_assists/src/auto_import.rs

```diff
@@ -0,0 +1,780 @@
+            }
+            (Some((segment, PATH_SEGMENT)), _, _) => {
+                segments.push(ast::PathSegment::cast(segment)?);
+                break;
+            }
+            (_, _, _) => return None,
+        }
+    }
+    // We need to reverse only the new added segments
+    let only_new_segments = segments.split_at_mut(oldlen).1;
+    only_new_segments.reverse();
+    return Some(segments.len() - oldlen);
+}
+
+fn fmt_segments(segments: &[&ast::PathSegment]) -> String {
+    let mut buf = String::new();
+    fmt_segments_raw(segments, &mut buf);
+    return buf;
+}
+
+fn fmt_segments_raw(segments: &[&ast::PathSegment], buf: &mut String) {
+    let mut first = true;
+    for s in segments {
+        if !first {
+            buf.push_str("::");
+        }
+        match s.kind() {
+            Some(ast::PathSegmentKind::Name(nameref)) => buf.push_str(nameref.text()),
+            Some(ast::PathSegmentKind::SelfKw) => buf.push_str("self"),
+            Some(ast::PathSegmentKind::SuperKw) => buf.push_str("super"),
+            Some(ast::PathSegmentKind::CrateKw) => buf.push_str("crate"),
+            None => {}
+        }
+        first = false;
+    }
+}
+
+#[derive(Copy, Clone)]
+enum PathSegmentsMatch {
+    // Patch matches exactly
+    Full,
+    // None of the segments matched. It's a more explicit Partial(0)
+    Empty,
+    // When some of the segments matched
+    Partial(usize),
+    // When all the segments of the right path are matched against the left path,
+    // but the left path is longer.
+    PartialLeft(usize),
+    // When all the segments of the left path are matched against the right path,
+    // but the right path is longer.
+    PartialRight(usize),
+    // In all the three cases above we keep track of how many segments matched
+}
+
+fn compare_path_segments(
+    left: &[&ast::PathSegment],
+    right: &[&ast::PathSegment],
+) -> PathSegmentsMatch {
+    let mut matching = 0;
+    for either_or_both in left.iter().zip_longest(right.iter()) {
+        match either_or_both {
+            EitherOrBoth::Both(left, right) => {
+                if compare_path_segment(left, right) {
+                    matching += 1
+                } else {
+                    return if matching == 0 {
+                        PathSegmentsMatch::Empty
+                    } else {
+                        PathSegmentsMatch::Partial(matching)
+                    };
+                }
+            }
+            EitherOrBoth::Left(_) => {
+                return PathSegmentsMatch::PartialLeft(matching);
+            }
+            EitherOrBoth::Right(_) => {
+                return PathSegmentsMatch::PartialRight(matching);
+            }
+        }
+    }
+    return PathSegmentsMatch::Full;
+}
+
+fn compare_path_segment(a: &ast::PathSegment, b: &ast::PathSegment) -> bool {
+    if let (Some(ka), Some(kb)) = (a.kind(), b.kind()) {
+        return match (ka, kb) {
```

Comment:
Don't need this return. That's actually also a warning we can already implement.